### PR TITLE
Feature/rpnv1 358 wall clock timestamp

### PR DESCRIPTION
**Pull request by clstrfsck**

### CODE REVIEW

**Comment by clstrfsck** [comment]
Uncovered lines are all in org.radix.api.* package.  I propose that we postpone increasing the coverage here until after the proposed API rewrite.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/MockedLedgerModule.java [hunk 1]
```diff
@@ -30,34 +32,38 @@
...
 		bind(CommittedStateSyncRx.class).toInstance(Observable::never);
 		bind(EpochChangeRx.class).toInstance(Observable::never);
-		EpochChange initialEpoch = new EpochChange(VertexMetadata.ofGenesisAncestor(validatorSet), validatorSet);
-		bind(EpochChange.class).toInstance(initialEpoch);
 		bind(NextCommandGenerator.class).toInstance((view, aids) -> null);
 		bind(SyncRequestSender.class).toInstance(req -> { });
 	}
 
+	@Provides
+	@Singleton
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/MockedLedgerModule.java:55]: Doesn't need to be singleton

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/DeterministicRandomlySyncedLedgerModule.java [hunk 2]
```diff
@@ -53,7 +52,7 @@ Ledger syncedExecutor(CommittedStateSyncSender committedStateSyncSender) {
 		return new Ledger() {
 			@Override
 			public PreparedCommand prepare(Vertex vertex) {
-				return PreparedCommand.create(0, Hash.ZERO_HASH);
+				return PreparedCommand.create(0, vertex.getQC().getTimestampedSignatures().weightedTimestamp());
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/DeterministicRandomlySyncedLedgerModule.java]: Given that this is a mocked Ledger would prefer to see this be 0L to emphasize/prove that consensus does not depend on the result of prepare().

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/MockedLedgerModule.java [hunk 3]
```diff
@@ -30,34 +32,38 @@
...
+	}
+
 	@Provides
 	@Singleton
 	Ledger syncedLedger() {
 		return new Ledger() {
 			@Override
 			public PreparedCommand prepare(Vertex vertex) {
-				return PreparedCommand.create(0, Hash.ZERO_HASH);
+				return PreparedCommand.create(0, vertex.getQC().getTimestampedSignatures().weightedTimestamp());
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/MockedLedgerModule.java]: Given that this is a mocked Ledger would prefer to see this be 0L to emphasize/prove that consensus does not depend on the result of prepare().

File: radixdlt/src/main/java/org/radix/api/http/RadixHttpServer.java [hunk 4]
```diff
@@ -282,6 +289,8 @@ private void addRestRoutesTo(RoutingHandler handler) {
 			respond(result, exchange);
 		}, handler);
 
+		addRoute("/api/bft", Methods.PUT_STRING, this::handleBftState, handler);
```
Comment (talekhinezh) [inline radixdlt/src/main/java/org/radix/api/http/RadixHttpServer.java]: What's the use case for this given we have a JSON-RPC for this already?

File: radixdlt/src/main/java/org/radix/api/http/RadixHttpServer.java [hunk 5]
```diff
@@ -362,6 +371,51 @@ public final JSONObject disconnectAllPeers() {
...
+	 *   <li><b>id</b> the ID of the BFT instance (must be {@code 0} for now)</li>
+	 *   <li><b>state</b> {@code true} to enable the specified instance id, otherwise
+	 *     the instance is disabled
+	 * </ul>
+	 *
+	 * @param exchange The {@link HttpServerExchange} to use
+	 * @throws IOException if an error occurs parsing form data
+	 */
+	private void handleBftState(HttpServerExchange exchange) throws IOException {
+		try (final FormDataParser formDataParser = this.formParserFactory.createParser(exchange)) {
```
Comment (talekhinezh) [inline radixdlt/src/main/java/org/radix/api/http/RadixHttpServer.java]: Given that all of our apis use a JSON body, I think we should stay consistent and not introduce a form-data body.

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/DeterministicAlwaysSyncedLedgerModule.java [hunk 6]
```diff
@@ -44,7 +43,7 @@ Ledger syncedExecutor() {
 		return new Ledger() {
 			@Override
 			public PreparedCommand prepare(Vertex vertex) {
-				return PreparedCommand.create(0, Hash.ZERO_HASH);
+				return PreparedCommand.create(0, vertex.getQC().getTimestampedSignatures().weightedTimestamp());
```
Comment (talekhinezh) [inline radixdlt/src/integration/java/com/radixdlt/integration/distributed/deterministic/DeterministicAlwaysSyncedLedgerModule.java]: Given that this is a mocked Ledger would prefer to see this be 0L to emphasize/prove that consensus does not depend on the result of prepare().

**Review by talekhinezh — CHANGES_REQUESTED** [review]

**Review by talekhinezh — APPROVED** [review]

File: radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/MockedLedgerModule.java (hunk 1 above)
Comment (clstrfsck) [reply radixdlt/src/integration/java/com/radixdlt/integration/distributed/simulation/MockedLedgerModule.java:55]: I think it does?  There are no guarantees here that the parameters are fixed.

File: radixdlt/src/main/java/org/radix/api/http/RadixHttpServer.java (hunk 4 above)
Comment (clstrfsck) [reply radixdlt/src/main/java/org/radix/api/http/RadixHttpServer.java]: To make it easier to make this inaccessible using nginx.  For the JSON RPC will need to statefully inspect the stream, which we probably don't want to do in a real system.

**Pull request merged**